File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    
    return binarySearch(merge(flattened(matrix)), target)
    
    // let arr = merge(flattened(matrix))
    // console.log(arr)
    
};

const binarySearch = (nums, target) => {

  let beg = 0
  let end = nums.length - 1

  let mid = Math.floor((beg + end) / 2)

  while(beg < end){

    if(nums[mid] == target){
      return true
    } else if( nums[mid] > target){
      end = mid - 1
    } else{
       beg = mid + 1
    }
      
    mid = Math.floor(((beg + end) / 2))

  }

  if(nums[mid] == target){
    return true
  }

  return false
  
}

const merge = (arr) =>{
  if(arr.length == 1){
    return arr
  }

  let mid = arr.length / 2
  let subLeft = merge(arr.slice(0, mid))
  let subRight = merge(arr.slice(mid))

  return mergeArr(subLeft, subRight)
}

const mergeArr = (subLeft, subRight) => {

  let res = []
  let l = 0
  let r = 0

  while(l < subLeft.length && r < subRight.length){

    if(subLeft[l] < subRight[r]){
      res.push(subLeft[l])
      l++
    } else {
      res.push(subRight[r])
      r++
    }
    
  }

  return  res.concat(subLeft.slice(l)).concat(subRight.slice(r))
}

const flattened = (arr, lis=[]) => {



    for(let i = 0; i < arr.length; i++){

      if(Array.isArray(arr[i])){
        flattened(arr[i], lis)
      }
      else{
        lis.push(arr[i])
      }

    }

    return lis

}
  

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
    
        if len(matrix) == 0:
            return -1
   
        first = 0
        last = len(matrix) - 1
        middle = (first + last) // 2

        while first <= last:

            if matrix[middle][0] > target :
                last = middle -1
            elif matrix[middle][len(matrix[middle]) - 1] < target:
                first = middle + 1
            else:
                if self.search(matrix[middle], target) != -1:
                    print("t")
                    return True
                else:
                    print("f")
                    return False
                
            middle = (first + last) // 2



        return False
        

        

    
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
   
        first = 0
        last = len(nums) - 1
        middle = (first + last) // 2

        while first <= last:

            if nums[middle] == target:
                return middle

            elif nums[middle] > target:
                last = middle -1
            elif nums[middle] < target:
                first = middle + 1
            middle = (first + last) // 2



        return -1
